src/docfx/lib/git/GitRepoInfoProvider.cs

```diff
@@ -1,17 +1,74 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System;
 using System.Collections.Concurrent;
+using System.Collections.Generic;
 using System.Diagnostics;
 using System.IO;
+using System.Linq;
 
 namespace Microsoft.Docs.Build
 {
     internal class GitRepoInfoProvider
     {
-        private readonly ConcurrentDictionary<string, GitRepoInfo> _cache
+        private static readonly string s_defaultCachePath = Path.Combine(
+            Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
+            ".docfx",
+            "cache",
+            "user-profile.json");
+
+        private readonly ConcurrentDictionary<string, GitRepoInfo> _folderRepoInfocache
             = new ConcurrentDictionary<string, GitRepoInfo>();
 
+        private readonly Lazy<Dictionary<string, List<GitCommit>>> _fileCommitsCache;
+
+        private readonly GitUserProfileCache _githubUserProfileCache;
+
+        public GitRepoInfoProvider(Docset docset)
+        {
+            Debug.Assert(docset != null);
+
+            _fileCommitsCache = new Lazy<Dictionary<string, List<GitCommit>>>(() => LoadCommits(docset));
+            var userProfileCachePath = string.IsNullOrEmpty(docset.Config.Contributor.UserProfileCachePath)
+                ? s_defaultCachePath
+                : Path.Combine(docset.DocsetPath, docset.Config.Contributor.UserProfileCachePath);
+            _githubUserProfileCache = GitUserProfileCache.Create(userProfileCachePath);
+        }
+
+        public (GitUserInfo author, GitUserInfo[] contributors, DateTime updatedAt) GetContributorInfo(Document document)
+        {
+            Debug.Assert(document != null);
+
+            // TODO: support specifed authorName and updatedAt
+            GitUserProfile authorInfo = null;
+            if (!_fileCommitsCache.Value.TryGetValue(document.FilePath, out var commits)
+                || commits.Count == 0)
+            {
+                return (null, null, DateTime.Now);
```

Comment:
Prefer Array.Empty<GitUserInfo> over null

Reply:
I think null can help template quickly determine that "contributors" section isn't needed for this page.

---
In reply to: [198002337](https://github.com/dotnet/docfx/pull/2912#discussion_r198002337) [](ancestors = 198002337)